src/shell.js

```diff
@@ -193,11 +193,35 @@ if (ENVIRONMENT_IS_NODE) {
   if (typeof process == 'undefined' || !process.release || process.release.name !== 'node') throw new Error('not compiled for this environment (did you build to HTML and try to run it not on the web, or set ENVIRONMENT to something - like node - and run it someplace else - like on the web?)');
 #endif
 #endif
+  // `require()` is no-op in an ESM module, use `createRequire()` to construct
+  // the require()` function.  This is only necessary for multi-environment
+  // builds, `-sENVIRONMENT=node` emits a static import declaration instead.
+  // TODO: Swap all `require()`'s with `import()`'s?
+#if EXPORT_ES6 && ENVIRONMENT_MAY_BE_WEB
+  const { createRequire } = await import('module');
+  /** @suppress{duplicate} */
+  var require = createRequire(import.meta.url);
+#endif
+  // These modules will usually be used on Node.js. Load them eagerly to avoid
+  // the complexity of lazy-loading.
+  var fs = require('fs');
+  var nodeUrl = require('url');
+  var nodePath = require('path');
+
   if (ENVIRONMENT_IS_WORKER) {
-    scriptDirectory = require('path').dirname(scriptDirectory) + '/';
+    scriptDirectory = nodePath.dirname(scriptDirectory);
   } else {
-    scriptDirectory = __dirname + '/';
+#if EXPORT_ES6
+    // EXPORT_ES6 + ENVIRONMENT_IS_NODE always requires use of import.meta.url,
+    // since there's no way getting the current absolute path of the module when
+    // support for that is not available.
+    scriptDirectory = nodePath.dirname(nodeUrl.fileURLToPath(import.meta.url));
```

Comment:
This one I'm less sure about, but also worth checking if we can get away with just keeping the URL here.

Reply:
Indeed, I wasn't sure about this. I removed the `fileURLToPath` construction with commit  cff5945db24ffc560421f28759efc2993f931925, however, that may cause the `locateFile` function to return `file://` strings on Node.js.
https://github.com/emscripten-core/emscripten/blob/84a634167a1cd9e8c47d37a559688153a4ceace6/src/shell.js#L152-L161
Let's see if the CI likes that, though locally it still passes the `other.test_emcc_output_mjs* other.test_export_es6*` tests.

Reply:
> however, that may cause the `locateFile` function to return `file://` strings on Node.js.

Now I'm really curious, if `file://` strings are not accepted directly, then how would same strings returned from `locateFile` work...

Reply:
> Now I'm really curious, if `file://` strings are not accepted directly, then how would same strings returned from `locateFile` work...

IIUC, `file://` strings are accepted after that `isFileURI(filename) ? new URL(filename) ...` change in `read_`, which will probably always be called on Node.js when reading files. But maybe it lacks some test coverage for this.

Reply:
Oh ok, that makes sense. (well, at least if nothing else uses the result of `locateFile`)

Reply:
I suppose it could also cause issues if someone does this:
```js
Module['locateFile'] = function (fileName, scriptDirectory) {
  if (scriptDirectory.startsWith('file://')) {
    throw new TypeError();
  }
  return scriptDirectory + fileName;
}
```
(i.e. overriding `Mode.locateFile` with custom logic based on `scriptDirectory`)

But I'm not sure if this is done in the wild.

Reply:
Maybe it's safer to keep scriptDirectory as actual filepath then, at least for now.

Reply:
9074bb71ac76b35057cb27ca03be5f864a3f9444 :)

Reply:
Though, while doing that, I noticed this:
https://github.com/emscripten-core/emscripten/blob/84a634167a1cd9e8c47d37a559688153a4ceace6/src/shell.js#L136-L137
... a couple of lines further ...
https://github.com/emscripten-core/emscripten/blob/84a634167a1cd9e8c47d37a559688153a4ceace6/src/shell.js#L364-L370
So, perhaps `scriptDirectory` was already a `file://`-string on Node.js when linking with `-sEXPORT_ES6`? Let me verify this.

Reply:
... nvm, I didn't look carefully, that is guarded with `#if SHARED_MEMORY && !MODULARIZE`, though `MODULARIZE` is forcefully enabled when linking with `-sEXPORT_ES6`:
https://github.com/emscripten-core/emscripten/blob/89a494fa1b1c6e10210d8e1e87dd9f475dd6b4a1/emcc.py#L2253-L2257
So I'm not sure if this comment is up-to-date:
https://github.com/emscripten-core/emscripten/blob/84a634167a1cd9e8c47d37a559688153a4ceace6/src/shell.js#L134-L135